# Machine Learning - CA2  
## Neural Networks and Semantic Analysis

**Student:** Thiago Gonçalves da Costa — **2022161**  
**College:** CCT College Dublin  
**Programme:** BSc (Hons) in Computing and IT — Year 4  
**Module:** Machine Learning for AI  
**Assessment:** ML_CA2  
**Lecturer:** David McQuaid  
**Submission Deadline:** 31st May 2026  

## Introduction

This project is divided into two main parts: **Neural Network Regression** and **Semantic / Sentiment Analysis**.

The first part uses the `international-visitors-london.csv` dataset. The main objective is to prepare and analyse the data, then build machine learning models to predict visitor spending in London. The target variable for this task is **Spend (£m)**. Because this target is a continuous numerical value, this part of the project is a supervised regression problem.

In this section, I will build a Neural Network model and compare it with at least one standard machine learning regression model. The models will be tested and evaluated to understand which one performs better for this dataset. A prediction will also be made for a new visitor segment that is not already in the original dataset.

The second part of the project uses social media text data to perform sentiment analysis. The text observations will be classified as **positive**, **neutral**, or **negative**. The overall sentiment distribution will then be visualised using Python.

All data preparation, analysis, modelling, evaluation, visualisation, and written discussion are completed in this Jupyter Notebook using Python and Markdown.

## Project Methodology


This project follows the general workflow covered in class: understanding the problem, understanding the data, preparing the data, building models, evaluating the results, and discussing the final findings (McQuaid, 2026o).

This structure helps to keep the notebook organised and makes sure that the data is explored and prepared before any modelling is done.

For the regression part, this means loading and understanding the London visitors dataset, cleaning and preparing the data, building regression models, comparing their results, and making a final prediction.

For the sentiment analysis part, this means preparing a social media text dataset, classifying the text as positive, neutral, or negative, visualising the sentiment distribution, and discussing the findings.

# Part 1 — Spend (£m) Prediction

## Business Understanding

The first part of this project focuses on predicting visitor spending in London using the `international-visitors-london.csv` dataset.

The main business problem is to understand and predict how much different international visitor segments spend in London. In the dataset, this is represented by the target variable **Spend (£m)**.

Visitor spending is important because it helps to understand the economic value of international tourism in London. If spending can be predicted from visitor characteristics, it can help tourism analysts and businesses understand which types of visitor segments are linked with higher or lower spending.

For example, spending may be affected by factors such as the visitor market, quarter of the year, duration of stay, mode of transport, purpose of visit, number of visits, and number of nights.

From a machine learning point of view, this is a **supervised regression problem**. It is supervised because the dataset already contains the target values, and it is regression because **Spend (£m)** is a continuous numerical value.

The machine learning objective is to build models that can predict **Spend (£m)** using the available visitor segment features. In this project, I will train a Neural Network model and compare it with at least one standard machine learning regression model, as required in the assessment brief.

An important point is that each row in this dataset does not represent one individual visitor. Instead, each row represents an aggregated visitor segment. For example, a row may describe visitors from a specific market, during a specific quarter, travelling for a specific purpose, using a specific mode of transport, and staying for a certain duration.

Because of this, when this notebook later makes a prediction for a "new visitor", it will be treated as a prediction for a new visitor segment, not for one individual person. This is more suitable for the structure of the dataset.

The success of the models will be evaluated by comparing predicted spending with the real spending values in the test set. Regression metrics such as MAE, MSE, RMSE, and R² will be used to compare the models and help decide which model performs better.

The final aim of this part is to prepare the data, build and test regression models, compare their results, select a suitable final model, and use it to predict the **Spend (£m)** for a new visitor segment.

## Import Libraries

Before loading and analysing the dataset, the Python libraries needed for this project are imported.

The main libraries used are:

- **pandas** and **numpy** for loading, handling and preparing the data;
- **matplotlib** and **seaborn** for data visualisation;
- **scikit-learn** for data preparation, standard machine learning regression models, train/test split, scaling and evaluation metrics;
- **TensorFlow / Keras** for building the Neural Network model later in the project.

At this stage, I am developing the first parts of the notebook on a Mac environment. The TensorFlow/Keras import is commented for now because this environment is not currently loading TensorFlow correctly. The Neural Network section will be run later in a Windows environment, where TensorFlow/Keras can be imported and executed properly.

This allows the data loading, data understanding, data preparation, exploratory analysis and standard regression modelling stages to continue without stopping the notebook (McQuaid, 2026g; McQuaid, 2026k; McQuaid, 2026a).

In [1]:
# Data handling
import pandas as pd
import numpy as np

# Data visualisation
import matplotlib.pyplot as plt
import seaborn as sns

# Data preparation and model evaluation
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Standard machine learning regression models
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor

# Neural Network model
# These imports are commented while developing on Mac.
# They will be uncommented and run later in a Windows environment.
# from tensorflow.keras.models import Sequential
# from tensorflow.keras.layers import Dense

# Display settings
pd.set_option("display.max_columns", None)

## Load Dataset

The dataset used for this part of the project is `international-visitors-london.csv`.

In this step, the dataset is loaded into a pandas DataFrame so it can be explored, cleaned, prepared and later used for modelling. This follows the class approach of first reading the data into a DataFrame before checking the structure, data types and summary statistics (McQuaid, 2026g).

The file is loaded using `latin1` encoding because the dataset contains the pound symbol in the target column **Spend (£m)**. Using this encoding avoids possible reading errors when loading the CSV file.

In [2]:
# Load the London visitors dataset
df = pd.read_csv("international-visitors-london.csv", encoding="latin1")

# Display the first five rows
df.head()

,year,quarter,market,dur_stay,mode,purpose,area,Visits (000s),Spend (£m),Nights (000s),sample
0,2002,January-March,Belgium,1-3 nights,Air,Holiday,LONDON,3.572186,0.969138,6.954456,5
1,2002,January-March,Belgium,1-3 nights,Air,Business,LONDON,9.284226,2.399577,12.604959,19
2,2002,January-March,Belgium,1-3 nights,Air,VFR,LONDON,0.877182,0.089833,2.153128,3
3,2002,January-March,Belgium,1-3 nights,Air,Miscellaneous,LONDON,0.163874,0.010160,0.163874,1
4,2002,January-March,Belgium,1-3 nights,Sea,Business,LONDON,1.648670,0.016789,1.650300,1


The first five rows were displayed to get an initial view of the dataset before starting the data understanding stage.

From the output, the dataset contains categorical columns such as year, quarter, market, duration of stay, mode, purpose, and area. It also contains numerical columns such as Visits (000s), Spend (£m), Nights (000s), and sample.

The target variable for this regression task is Spend (£m).

## Data Understanding

In this section, I will explore the structure of the London visitors dataset before doing any cleaning, preparation or modelling.

This step is important because the dataset needs to be understood before choosing the correct preparation steps and models. In the class materials, the data understanding stage includes describing the data, exploring the data and checking its quality before moving to data preparation and modelling (McQuaid, 2026o).

The main checks in this section are:

- the number of rows and columns;
- the column names;
- the data types;
- numerical summary statistics;
- categorical summary statistics;
- the number of unique values in each column;
- the distribution of the main categorical variables;
- the target variable for the regression task.

### Dataset Shape

First, I will check the shape of the dataset. This shows how many rows and columns are available in the DataFrame.

Checking the shape is useful because it gives an initial idea of the size of the dataset before starting data preparation or modelling.

In [3]:
# Check the number of rows and columns
df.shape

(61462, 11)

The dataset shape shows that there are **61,462 rows** and **11 columns**.

This means the dataset has a good number of observations for a regression task. Each row represents an aggregated visitor segment, and each column represents information about that segment.

### Column Names

Next, I will check the column names in the dataset. This helps to identify the available features and the target variable.

In [4]:
# Display the column names
df.columns

Index(['year', 'quarter', 'market', 'dur_stay', 'mode', 'purpose', 'area',
       'Visits (000s)', 'Spend (£m)', 'Nights (000s)', 'sample'],
      dtype='object')

The dataset contains columns related to the time period, visitor market, stay duration, travel mode, purpose, area, visits, spend, nights and sample size.

The target variable for this part of the project is **Spend (£m)** because the objective is to predict visitor spending.

### Dataset Information

Next, I will use `info()` to check the data types and the number of non-null values in each column.

This follows the class approach of using pandas to inspect the structure of the DataFrame before modelling (McQuaid, 2026g).

In [5]:
# Check data types and non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61462 entries, 0 to 61461
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   year           61462 non-null  object 
 1   quarter        61462 non-null  object 
 2   market         61462 non-null  object 
 3   dur_stay       61462 non-null  object 
 4   mode           61462 non-null  object 
 5   purpose        61462 non-null  object 
 6   area           61462 non-null  object 
 7   Visits (000s)  61462 non-null  float64
 8   Spend (£m)     61462 non-null  float64
 9   Nights (000s)  61462 non-null  float64
 10  sample         61462 non-null  int64  
dtypes: float64(3), int64(1), object(7)
memory usage: 5.2+ MB


The `info()` output shows that the dataset has **61,462 entries** and **11 columns**.

The dataset contains both categorical and numerical variables. The categorical columns are stored as object data types, while the numerical columns are stored as float or integer data types.

The output also shows that all columns have 61,462 non-null values. This means that, at this initial stage, there are no missing values visible in the dataset.

### Data Types

I will now display only the data types. This makes it easier to separate categorical columns from numerical columns.

In [6]:
# Display data types
df.dtypes

year              object
quarter           object
market            object
dur_stay          object
mode              object
purpose           object
area              object
Visits (000s)    float64
Spend (£m)       float64
Nights (000s)    float64
sample             int64
dtype: object

The data types show that the dataset contains object columns and numerical columns.

The object columns are mainly categorical variables, such as year, quarter, market, duration of stay, mode, purpose, and area.

The numerical columns are Visits (000s), Spend (£m), Nights (000s), and sample.

The year column is stored as an object instead of an integer. This will need to be checked later because the dataset contains a value called 2020P, which is not a normal numeric year.

### Numerical Summary Statistics

Next, I will use `describe()` to summarise the numerical columns.

This gives the count, mean, standard deviation, minimum, quartiles and maximum values for the numerical variables. This helps to understand the scale of the numeric features and the target variable before modelling.

In [7]:
# Summary statistics for numerical columns
df.describe()

,Visits (000s),Spend (£m),Nights (000s),sample
count,61462.000000,61462.000000,61462.000000,61462.000000
mean,4.927651,3.086634,30.300852,6.099021
std,10.061362,7.243874,61.898302,13.335820
min,0.000103,0.000000,0.000309,1.000000
25%,0.822888,0.321693,3.901581,1.000000
50%,1.797305,1.047351,10.676935,2.000000
75%,4.416618,3.007542,29.483044,5.000000
max,187.487282,373.232590,2492.178513,299.000000


The numerical summary shows the main statistics for Visits (000s), Spend (£m), Nights (000s), and sample.

The target variable Spend (£m) has a much higher maximum value than its median value. This suggests that the spending data may be right-skewed, with many lower values and some very high values.

The same pattern can also be seen in variables such as Visits (000s) and Nights (000s), where the maximum values are much larger than the median values. This will be explored more during the visualisation stage.

### Categorical Summary Statistics

The normal `describe()` function mainly summarises numerical columns. For categorical columns, I will use `describe(include='object')`.

This follows the pandas class material, where `describe(include=object)` is used to summarise categorical variables separately from numerical variables (McQuaid, 2026g).

In [8]:
# Summary statistics for categorical columns
df.describe(include="object")

,year,quarter,market,dur_stay,mode,purpose,area
count,61462,61462,61462,61462,61462,61462,61462
unique,19,4,62,4,3,5,1
top,2010,July-September,France,1-3 nights,Air,Holiday,LONDON
freq,3601,16827,2611,19710,46778,19695,61462


The categorical summary shows the number of unique values, the most frequent value, and the frequency of the most common value for each object column.

The area column has only one unique value, LONDON, and it appears in all 61,462 rows. This suggests that the column may not add useful information for prediction because all rows refer to the same area.

The market column has many different categories, which means it will need to be encoded before modelling. The quarter, duration of stay, mode, and purpose columns are also categorical and will need preparation before being used in machine learning models.

### Number of Unique Values

I will now check how many unique values exist in each column.

This helps to understand which variables are constant, which variables have only a few categories, and which variables have many categories.

In [9]:
# Count unique values in each column
df.nunique()

year                19
quarter              4
market              62
dur_stay             4
mode                 3
purpose              5
area                 1
Visits (000s)    55965
Spend (£m)       60614
Nights (000s)    60368
sample             195
dtype: int64

The number of unique values gives a clearer view of the dataset structure.

The numerical columns have many unique values, which is expected because they contain continuous or near-continuous measurements such as visits, spend and nights.

The area column has only one unique value, so it may not be useful for prediction. The market column has many unique values, which means it contains many visitor markets and will need careful encoding later.

The year column has 19 unique values because the dataset includes years from 2002 up to 2020P. The value 2020P will need to be handled during data preparation.

### Value Counts for Main Categorical Columns

Next, I will check the distribution of the main categorical columns using `value_counts()`.

This helps to understand how the observations are distributed across years, quarters, markets, duration of stay, mode of transport, purpose and area.

In [10]:
# Check value counts for the main categorical columns
categorical_columns = ["year", "quarter", "market", "dur_stay", "mode", "purpose", "area"]

for col in categorical_columns:
    print(f"\nValue counts for {col}:")
    print(df[col].value_counts())


Value counts for year:
year
2010     3601
2013     3571
2009     3558
2012     3544
2004     3521
2005     3494
2011     3485
2006     3471
2015     3470
2014     3466
2016     3402
2002     3397
2017     3295
2007     3290
2003     3240
2008     3094
2019     2980
2018     2953
2020P     630
Name: count, dtype: int64

Value counts for quarter:
quarter
July-September      16827
April-June          15336
January-March       14720
October-December    14579
Name: count, dtype: int64

Value counts for market:
market
France         2611
USA            2498
Germany        2206
Netherlands    1957
Australia      1943
               ... 
Bahrain         268
Indonesia       254
Chile           242
Oman            219
Serbia          171
Name: count, Length: 62, dtype: int64

Value counts for dur_stay:
dur_stay
1-3  nights    19710
4-7  nights    17525
8-14 nights    12819
15+  nights    11408
Name: count, dtype: int64

Value counts for mode:
mode
Air       46778
Tunnel     8682
Sea        6002

The value counts show how the observations are distributed across the categorical columns.

The year column shows data from 2002 to 2020P. Most years have around 3,000 to 3,600 rows, but 2020P only has 630 rows. This suggests that 2020P may be a provisional or partial year, so it will need to be handled carefully during preparation.

The quarter column contains four time periods, and the observations are reasonably spread across them. July-September has the highest number of rows.

The market column contains 62 different visitor markets. France, USA and Germany appear among the most common markets in the dataset. This column may be important because visitor spending can vary depending on where visitors come from.

The dur_stay column contains four stay duration groups. Some values include extra spaces between the number range and the word nights, so this column will need string standardisation during data cleaning.

The mode column shows that Air is the most common mode of transport. The purpose column shows that Holiday, VFR and Business are the most common travel purposes.

The area column only contains LONDON, so it will be reviewed later during data preparation because it may not provide predictive value.

### Identifying Numerical and Categorical Columns

Before moving to data cleaning and preparation, I will separate the numerical and categorical columns. This makes the next steps easier because numerical and categorical variables usually need different preparation methods.

In [11]:
# Identify numerical and categorical columns
numerical_columns = df.select_dtypes(include=["int64", "float64"]).columns
categorical_columns = df.select_dtypes(include=["object"]).columns

print("Numerical columns:")
print(list(numerical_columns))

print("\nCategorical columns:")
print(list(categorical_columns))

Numerical columns:
['Visits (000s)', 'Spend (£m)', 'Nights (000s)', 'sample']

Categorical columns:
['year', 'quarter', 'market', 'dur_stay', 'mode', 'purpose', 'area']


The numerical columns are Visits (000s), Spend (£m), Nights (000s), and sample.

The categorical columns are year, quarter, market, dur_stay, mode, purpose, and area.

This confirms that the dataset contains a mix of numerical and categorical data. The categorical variables will need to be converted into numeric format before being used in machine learning models.

### Target Variable Check

Finally, I will check the target variable **Spend (£m)** separately.

This is important because Spend (£m) is the value that the regression models and Neural Network will try to predict.

In [12]:
# Check the target variable
df["Spend (£m)"].describe()

count    61462.000000
mean         3.086634
std          7.243874
min          0.000000
25%          0.321693
50%          1.047351
75%          3.007542
max        373.232590
Name: Spend (£m), dtype: float64

The target variable **Spend (£m)** is numerical and continuous, which confirms that this is a regression task.

The summary shows that the target has many lower values and some very high values. This suggests that the target may contain outliers or a right-skewed distribution. This will be explored further during the visualisation stage before modelling.

### Data Understanding Summary

From the initial data understanding stage, the dataset contains **61,462 rows** and **11 columns**.

The target variable is **Spend (£m)**, which is a continuous numerical variable. This confirms that the first part of the project is a supervised regression problem.

The dataset contains a mix of categorical and numerical variables. The main categorical variables are **year**, **quarter**, **market**, **duration of stay**, **mode**, **purpose**, and **area**. The main numerical variables are **Visits (000s)**, **Spend (£m)**, **Nights (000s)**, and **sample**.

Each row represents an aggregated visitor segment rather than one individual visitor. This is important for interpreting the model later, because the final prediction will be for a new visitor segment.

The initial checks also show some points that will need attention in the data preparation stage:

- the **year** column contains **2020P**, so it is stored as text;
- the **area** column contains only **LONDON**, so it may not add predictive value;
- the **dur_stay** column contains some extra spacing in the category names;
- categorical variables will need to be encoded before modelling;
- numerical variables may need scaling before training the Neural Network;
- **Spend (£m)** appears to be right-skewed and may contain outliers.

The next stage will be data quality checking and data preparation, where these issues will be investigated and handled before building the models.